tests/Test.ADAL.NET.Unit/SovereignCloudsTests.cs

```diff
@@ -0,0 +1,82 @@
+            platformParameters = new PlatformParameters(PromptBehavior.Auto);
+        }
+
+        [TestMethod]
+        [Description("Sovereign user use world wide authority")]
+        public async Task SovereignUserWorldWideAuthorityIntegrationTest()
+        {
+            const string testCloudInstanceName = "some-sovereign-cloud";
+            const string sovereignAuthorityHostPrefix = "login.";
+            const string sovereignAuthorityHost = sovereignAuthorityHostPrefix + testCloudInstanceName;
+
+            var sovereignAuthority =
+                new UriBuilder(TestConstants.DefaultAuthorityCommonTenant) {Host = sovereignAuthorityHost}.Uri
+                    .ToString();
+
+            // creating AuthenticationContext with common Authority
+            var authenticationContext =
+                new AuthenticationContext(TestConstants.DefaultAuthorityCommonTenant, false, new TokenCache());
+
+            // mock value for authentication returnedUriInput, with cloud_instance_name claim
+            var authReturnedUriInputMock = TestConstants.DefaultRedirectUri + "?code=some-code" + "&" +
+                                           TokenResponseClaim.CloudInstanceName + "=" + testCloudInstanceName;
+
+            MockHelpers.ConfigureMockWebUI(
+                new AuthorizationResult(AuthorizationStatus.Success, authReturnedUriInputMock),
+                // validate that authorizationUri passed to WebUi contains instance_aware query parameter
+                new Dictionary<string, string> {{"instance_aware", "true"}});
+
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler
+            {
+                Method = HttpMethod.Post,
+                ResponseMessage = new HttpResponseMessage(HttpStatusCode.OK)
+                {
+                    Content = new StringContent(
+                        "{\"token_type\":\"Bearer\",\"expires_in\":\"3599\",\"access_token\":\"some-access-token\"}")
+                },
+
+                AdditionalRequestValidation = request =>
+                {
+                    // make sure that Sovereign authority was used for Authorization request
+                    Assert.AreEqual(sovereignAuthorityHost, request.RequestUri.Authority);
+                }
+            });
+
+            var authenticationResult = await authenticationContext.AcquireTokenAsync(TestConstants.DefaultResource,
+                TestConstants.DefaultClientId,
+                TestConstants.DefaultRedirectUri, platformParameters, UserIdentifier.AnyUser, "instance_aware=true");
+
+            // make sure that sovereign Authority returned to the app in AuthenticationResult
+            Assert.AreEqual(sovereignAuthorityHost, new Uri(authenticationResult.Authority).Host);
+
+            // make sure that AuthenticationContext Authority was not changed
+            Assert.AreEqual(TestConstants.DefaultAuthorityCommonTenant, authenticationContext.Authority);
```

Comment:
we should not return authority with "common" as a tenant name. It should be the tenantid returned from the server. Same logic is already used for adding entry in token cache